File: api/tracelinks_api.py
"""
Модуль: tracelinks_api.py
Назначение: REST API управления связями трассировки
Автор: Разработчик
Дата создания: 21.03.2026
Требования: Interface.Software.TraceLinks, FR-05, FR-06, FR-12, FR-14
"""

from flask import Blueprint, jsonify, request
from flask_login import current_user, login_required

from models import Requirement, TraceLink
from services.traceability_service import (
    TraceLinkError,
    create_trace_link,
    delete_trace_link,
    get_links_for_requirement,
    run_full_integrity_check,
)

trace_api = Blueprint("tracelinks_api", __name__, url_prefix="/api/projects/<int:project_id>/tracelinks")


def link_to_dict(link):
    return {
        "id": link.id,
        "source_req_id": link.source_req_id,
        "source_system_id": link.source_req.system_id if link.source_req else None,
        "source_title": link.source_req.title if link.source_req else None,
        "target_req_id": link.target_req_id,
        "target_system_id": link.target_req.system_id if link.target_req else None,
        "target_title": link.target_req.title if link.target_req else None,
        "link_type": link.link_type,
        "description": link.description,
        "status": link.status,
        "created_at": link.created_at.isoformat() if link.created_at else None,
        "created_by": link.creator.username if link.creator else None,
    }


@trace_api.route("", methods=["GET"])
@login_required
def list_links(project_id):
    """Назначение: Список связей трассировки проекта."""
    links = TraceLink.query.join(
        Requirement, TraceLink.source_req_id == Requirement.id
    ).filter(
        Requirement.project_id == project_id
    ).order_by(TraceLink.created_at.desc()).all()

    return jsonify([link_to_dict(lnk) for lnk in links])


@trace_api.route("", methods=["POST"])
@login_required
def create_link(project_id):
    """
    Назначение: Создание связи трассировки
    Id требования: LLR_TraceLinkService_CreateLink_01-02
    """
    if not current_user.can_manage_links():
        return jsonify({"error": "Недостаточно прав"}), 403

    data = request.get_json(silent=True) or {}
    try:
        link = create_trace_link(
            source_id=data.get("source_req_id"),
            target_id=data.get("target_req_id"),
            link_type=data.get("link_type", "derives_from"),
            user_id=current_user.id,
            description=data.get("description", ""),
        )
        return jsonify(link_to_dict(link)), 201
    except TraceLinkError as exc:
        return jsonify({"error": exc.message, "code": exc.code}), 400


@trace_api.route("/<int:link_id>", methods=["DELETE"])
@login_required
def delete_link(project_id, link_id):
    """Назначение: Удаление связи трассировки."""
    if not current_user.can_manage_links():
        return jsonify({"error": "Недостаточно прав"}), 403

    try:
        delete_trace_link(link_id, current_user.id)
        return jsonify({"message": "Удалено"})
    except TraceLinkError as exc:
        return jsonify({"error": exc.message, "code": exc.code}), 400


@trace_api.route("/requirement/<int:req_id>", methods=["GET"])
@login_required
def links_for_req(project_id, req_id):
    """Назначение: Связи для конкретного требования."""
    links = get_links_for_requirement(req_id)
    return jsonify([link_to_dict(lnk) for lnk in links])


@trace_api.route("/integrity_check", methods=["POST"])
@login_required
def integrity_check(project_id):
    """
    Назначение: Полная проверка целостности
    Id требования: Functional.Traceability.IntegrityCheck.OnDemand
    """
    if not current_user.can_run_integrity_check():
        return jsonify({"error": "Недостаточно прав"}), 403

    issues = run_full_integrity_check(project_id, current_user.id)
    return jsonify({
        "issues_count": len(issues),
        "issues": [
            {
                "link_id": item["link"].id,
                "reason": item["reason"],
            }
            for item in issues
        ],
    })

File: backend/src/api/tracelinks_api.cpp
/*
Модуль: tracelinks_api.cpp
Назначение: Реализация REST API связей трассировки между требованиями
Автор: Разработчик
Дата создания: 21.03.2026
Изменения: 26.03.2026 — добавлена обработка исключений (try/catch) во всех обработчиках
Требования: Interface.Software.TraceLinks, FR-05, FR-06
*/

#include "tracelinks_api.hpp"
#include "middleware.hpp"
#include "services/traceability_service.hpp"

/*
Назначение: Преобразование объекта TraceLink в JSON
Входные данные: l — объект связи трассировки
Выходные данные: JSON представление связи
*/
static json link_to_json(const TraceLink& l)
{
    return {
        {"id", l.id},
        {"source_req_id", l.source_req_id},
        {"source_system_id", l.source_system_id},
        {"source_title", l.source_title},
        {"target_req_id", l.target_req_id},
        {"target_system_id", l.target_system_id},
        {"target_title", l.target_title},
        {"link_type", l.link_type},
        {"description", l.description},
        {"status", l.status},
        {"created_at", l.created_at},
        {"created_by", l.creator_username.empty()
            ? json(nullptr) : json(l.creator_username)}
    };
}

/*
Назначение: Обработка GET /api/projects/:pid/tracelinks — список связей проекта
Входные данные: путь с id проекта
Выходные данные: JSON массив связей
*/
static void handle_list_links(const httplib::Request& req, httplib::Response& res)
{
    auto user = require_auth(req, res);
    if (!user) return;

    int pid = std::stoi(req.matches[1]);
    auto links = Database::instance().get_links_for_project(pid);
    json arr = json::array();
    for (const auto& l : links) {
        arr.push_back(link_to_json(l));
    }
    json_response(res, arr);
}

/*
Назначение: Обработка POST /api/projects/:pid/tracelinks — создание связи
Входные данные: JSON {source_req_id, target_req_id, link_type?, description?}
Выходные данные: JSON объект связи (201)
*/
static void handle_create_link(const httplib::Request& req, httplib::Response& res)
{
    auto user = require_auth(req, res);
    if (!user) return;
    if (!user->can_manage_links()) {
        json_error(res, "Недостаточно прав", 403);
        return;
    }

    auto body = parse_json_body(req);
    try {
        auto link = create_trace_link(
            body.value("source_req_id", 0),
            body.value("target_req_id", 0),
            body.value("link_type", "derives_from"),
            user->id, body.value("description", ""));
        json_response(res, link_to_json(link), 201);
    } catch (const TraceLinkError& e) {
        json_response(res, {{"error", e.what()}, {"code", e.code}}, 400);
    }
}

/*
Назначение: Обработка DELETE /api/projects/:pid/tracelinks/:lid — удаление связи
Входные данные: путь с id связи
Выходные данные: JSON {message}
*/
static void handle_delete_link(const httplib::Request& req, httplib::Response& res)
{
    auto user = require_auth(req, res);
    if (!user) return;
    if (!user->can_manage_links()) {
        json_error(res, "Недостаточно прав", 403);
        return;
    }

    int lid = std::stoi(req.matches[2]);
    try {
        delete_trace_link_svc(lid, user->id);
        json_response(res, {{"message", "Удалено"}});
    } catch (const TraceLinkError& e) {
        json_response(res, {{"error", e.what()}, {"code", e.code}}, 400);
    }
}

/*
Назначение: Обработка GET .../tracelinks/requirement/:rid — связи требования
Входные данные: путь с id требования
Выходные данные: JSON массив связей
*/
static void handle_links_for_requirement(const httplib::Request& req, httplib::Response& res)
{
    auto user = require_auth(req, res);
    if (!user) return;

    int rid = std::stoi(req.matches[2]);
    auto links = Database::instance().get_links_for_requirement(rid);
    json arr = json::array();
    for (const auto& l : links) {
        arr.push_back(link_to_json(l));
    }
    json_response(res, arr);
}

/*
Назначение: Обработка POST .../tracelinks/integrity_check — проверка целостности
Входные данные: путь с id проекта
Выходные данные: JSON {issues_count, issues}
*/
static void handle_integrity_check(const httplib::Request& req, httplib::Response& res)
{
    auto user = require_auth(req, res);
    if (!user) return;
    if (!user->can_run_integrity_check()) {
        json_error(res, "Недостаточно прав", 403);
        return;
    }

    int pid = std::stoi(req.matches[1]);
    auto issues = run_full_integrity_check(pid, user->id);
    json arr = json::array();
    for (const auto& issue : issues) {
        arr.push_back(issue);
    }
    json_response(res, json{
        {"issues_count", static_cast<int>(issues.size())},
        {"issues", arr}});
}

void register_tracelink_routes(httplib::Server& svr)
{
    svr.Get(R"(/api/projects/(\d+)/tracelinks)",
        [](const httplib::Request& req, httplib::Response& res)
    {
        try {
            handle_list_links(req, res);
        } catch (const std::exception& e) {
            json_error(res, e.what(), 500);
        }
    });

    svr.Post(R"(/api/projects/(\d+)/tracelinks)",
        [](const httplib::Request& req, httplib::Response& res)
    {
        try {
            handle_create_link(req, res);
        } catch (const std::exception& e) {
            json_error(res, e.what(), 500);
        }
    });

    svr.Delete(R"(/api/projects/(\d+)/tracelinks/(\d+))",
        [](const httplib::Request& req, httplib::Response& res)
    {
        try {
            handle_delete_link(req, res);
        } catch (const std::exception& e) {
            json_error(res, e.what(), 500);
        }
    });

    svr.Get(R"(/api/projects/(\d+)/tracelinks/requirement/(\d+))",
        [](const httplib::Request& req, httplib::Response& res)
    {
        try {
            handle_links_for_requirement(req, res);
        } catch (const std::exception& e) {
            json_error(res, e.what(), 500);
        }
    });

    svr.Post(R"(/api/projects/(\d+)/tracelinks/integrity_check)",
        [](const httplib::Request& req, httplib::Response& res)
    {
        try {
            handle_integrity_check(req, res);
        } catch (const std::exception& e) {
            json_error(res, e.what(), 500);
        }
    });
}
